File: spec/check_example_overlap.py
#!/usr/bin/env python3
"""Verify that runnable text-fence mappings share salient source identifiers."""

from pathlib import Path
import re
import sys


STOP = set(
    "fn let mut pub data enum trait impl type use as for in if else match while "
    "return break continue pass true false nil self Self where reified annotate "
    "shape and or not is string bool void i8 i16 i32 i64 u8 u16 u32 u64 f32 "
    "f64 list map Result Ok Err Any".split()
)


def identifiers(text: str) -> set[str]:
    return {
        value
        for value in re.findall(r"[^\W\d]\w*", text, re.UNICODE)
        if len(value) >= 3 and value not in STOP
    }


def salient_identifiers(text: str) -> set[str]:
    """Names that characterize an example rather than incidental prose/locals."""

    source = re.sub(r'"(?:\\.|[^"\\])*"', '""', text)
    source = re.sub(r"#.*", "", source)

    declared = set(
        re.findall(
            r"\b(?:fn|data|enum|trait|impl|type)\s+(?:\[[^]]+\]\s*)?([^\W\d]\w*)",
            source,
            re.UNICODE,
        )
    )
    members = set(re.findall(r"\.([^\W\d]\w*)\s*(?:\[|\()", source, re.UNICODE))
    associated = set(re.findall(r"::([^\W\d]\w*)\s*\(", source, re.UNICODE))
    capitals = {
        value
        for value in re.findall(r"\b[A-Z][A-Za-z0-9_]*\b", source)
        if value not in STOP
    }
    result = (declared | members | associated | capitals) & identifiers(source)
    return result or identifiers(source)


def text_blocks(path: Path) -> list[str]:
    return [
        match.group(1)
        for match in re.finditer(
            r"^```text\s*\n(.*?)^```", path.read_text(encoding="utf-8"), re.M | re.S
        )
    ]


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: check_example_overlap.py SPEC_DIR EXAMPLES_TSV", file=sys.stderr)
        return 2
    spec_dir = Path(sys.argv[1])
    examples = Path(sys.argv[2])
    failures: list[str] = []
    cache: dict[str, list[str]] = {}
    for row in examples.read_text(encoding="utf-8").splitlines()[1:]:
        specification, block_text, classification, fixture_text = row.split("\t")
        if classification not in {"accept", "mixed"}:
            continue
        blocks = cache.setdefault(
            specification, text_blocks(spec_dir / specification)
        )
        block = blocks[int(block_text) - 1]
        block_ids = salient_identifiers(block)
        if not block_ids:
            continue
        fixture_ids: set[str] = set()
        for fixture in fixture_text.split("|"):
            fixture_ids.update(
                identifiers(
                    (spec_dir / "conformance" / fixture).read_text(encoding="utf-8")
                )
            )
        overlap = block_ids & fixture_ids
        coverage = len(overlap) / len(block_ids)
        if coverage < 0.60:
            failures.append(
                f"{specification} block {block_text} covers {len(overlap)}/"
                f"{len(block_ids)} salient identifiers ({coverage:.0%}) with {fixture_text}; "
                f"missing {', '.join(sorted(block_ids - overlap))}"
            )
    if failures:
        print("example inventory overlap failures:", file=sys.stderr)
        for failure in failures:
            print(f"- {failure}", file=sys.stderr)
        return 1
    print("example inventory overlap passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: spec/check-example-overlap.ts
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";
import { fileURLToPath } from "node:url";

const stop = new Set(
  "fn let mut pub data enum trait impl type use as for in if else match while return break continue pass true false nil self Self where reified annotate shape and or not is string bool void i8 i16 i32 i64 u8 u16 u32 u64 f32 f64 list map Result Ok Err Any".split(
    " ",
  ),
);
const identifierPattern = /[^\W\d]\w*/gu;

function identifiers(text: string): Set<string> {
  return new Set(
    [...text.matchAll(identifierPattern)]
      .map(([value]) => value)
      .filter((value) => value.length >= 3 && !stop.has(value)),
  );
}

function salientIdentifiers(text: string): Set<string> {
  const source = text.replaceAll(/"(?:\\.|[^"\\])*"/g, '""').replaceAll(/#.*$/gm, "");
  const declared = new Set(
    [
      ...source.matchAll(/\b(?:fn|data|enum|trait|impl|type)\s+(?:\[[^\]]+\]\s*)?([^\W\d]\w*)/gu),
    ].map((match) => match[1]!),
  );
  const members = new Set(
    [...source.matchAll(/\.([^\W\d]\w*)\s*(?:\[|\()/gu)].map((match) => match[1]!),
  );
  const associated = new Set(
    [...source.matchAll(/::([^\W\d]\w*)\s*\(/gu)].map((match) => match[1]!),
  );
  const capitals = new Set(
    [...source.matchAll(/\b[A-Z][A-Za-z0-9_]*\b/g)]
      .map(([value]) => value)
      .filter((value) => !stop.has(value)),
  );
  const all = identifiers(source);
  const result = new Set(
    [...declared, ...members, ...associated, ...capitals].filter((value) => all.has(value)),
  );
  return result.size > 0 ? result : all;
}

function textBlocks(text: string): string[] {
  return [...text.matchAll(/^```text\s*\n(.*?)^```/gms)].map((match) => match[1]!);
}

async function main(args: readonly string[]): Promise<number> {
  if (args.length !== 2) {
    console.error("usage: check-example-overlap.ts SPEC_DIR EXAMPLES_TSV");
    return 2;
  }
  const specificationDirectory = resolve(args[0]!);
  const examplesPath = resolve(args[1]!);
  const failures: string[] = [];
  const blockCache = new Map<string, string[]>();
  const fixtureCache = new Map<string, string>();
  const rows = (await readFile(examplesPath, "utf8")).split(/\r?\n/).slice(1).filter(Boolean);
  for (const row of rows) {
    const [specification = "", blockText = "", classification = "", fixtureText = ""] =
      row.split("\t");
    if (!new Set(["accept", "mixed"]).has(classification)) continue;
    let blocks = blockCache.get(specification);
    if (!blocks) {
      blocks = textBlocks(await readFile(resolve(specificationDirectory, specification), "utf8"));
      blockCache.set(specification, blocks);
    }
    const blockIds = salientIdentifiers(blocks[Number(blockText) - 1]!);
    if (blockIds.size === 0) continue;
    const fixtureIds = new Set<string>();
    for (const fixture of fixtureText.split("|")) {
      const path = resolve(specificationDirectory, "conformance", fixture);
      let source = fixtureCache.get(path);
      if (source === undefined) {
        source = await readFile(path, "utf8");
        fixtureCache.set(path, source);
      }
      for (const value of identifiers(source)) fixtureIds.add(value);
    }
    const overlap = [...blockIds].filter((value) => fixtureIds.has(value));
    const coverage = overlap.length / blockIds.size;
    if (coverage < 0.6) {
      const missing = [...blockIds]
        .filter((value) => !fixtureIds.has(value))
        .sort()
        .join(", ");
      failures.push(
        `${specification} block ${blockText} covers ${overlap.length}/${blockIds.size} salient identifiers (${Math.round(coverage * 100)}%) with ${fixtureText}; missing ${missing}`,
      );
    }
  }
  if (failures.length > 0) {
    console.error(
      `example inventory overlap failures:\n${failures.map((failure) => `- ${failure}`).join("\n")}`,
    );
    return 1;
  }
  console.log("example inventory overlap passed");
  return 0;
}

const invokedPath = process.argv[1] ? resolve(process.argv[1]) : "";
if (invokedPath === fileURLToPath(import.meta.url))
  process.exitCode = await main(process.argv.slice(2));
